include/exec/system_scheduler.hpp

```diff
@@ -0,0 +1,663 @@
+/*
+ * Copyright (c) 2023 Lee Howes
+ *
+ * Licensed under the Apache License Version 2.0 with LLVM Exceptions
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *   https://llvm.org/LICENSE.txt
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "stdexec/execution.hpp"
+// For the default implementation, test will override
+#include "exec/static_thread_pool.hpp"
+
+#include <iostream>
+
+struct __exec_system_scheduler_interface;
+struct __exec_system_sender_interface;
+struct __exec_system_scheduler_impl;
+struct __exec_system_sender_impl;
+
+
+// Virtual interfaces to underlying implementations for initial simplicit
+// TODO: Potentially move to custom vtable implementations
+struct __exec_system_context_interface {
+  virtual __exec_system_scheduler_interface* get_scheduler() noexcept = 0;
+};
+
+// bulk function for scheduler to transmit from, will wrap actual function stub stored in real type
+using __exec_system_bulk_shape = long;
+using __exec_system_bulk_fn = void(void*, __exec_system_bulk_shape);
+struct __exec_system_bulk_function_object {
+  void* fn_state = nullptr;
+  __exec_system_bulk_fn* fn = nullptr;
+};
+
+
+struct __exec_system_scheduler_interface {
+  virtual stdexec::forward_progress_guarantee get_forward_progress_guarantee() const = 0;
+  virtual __exec_system_sender_interface* schedule() = 0;
+  // TODO: Move chaining in here to support chaining after a system_sender or other system_bulk_sender
+  // or don't do anything that specific?
+  virtual __exec_system_sender_interface* bulk(__exec_system_bulk_shape shp, __exec_system_bulk_function_object fn) = 0;
+  virtual bool equals(const __exec_system_scheduler_interface* rhs) const = 0;
+};
+
+struct __exec_system_operation_state_interface {
+  virtual void start() noexcept = 0;
+};
+
+struct __exec_system_receiver {
+  void* cpp_recv_ = nullptr;
+  void (*set_value)(void* cpp_recv);
+  void (*set_stopped)(void* cpp_recv);
+  // Type-erase the exception pointer for extern-c-ness
+  void (*set_error)(void* cpp_recv, void* exception);
+
+};
+
+struct __exec_system_sender_interface {
+  virtual __exec_system_operation_state_interface* connect(__exec_system_receiver recv) noexcept = 0;
+  virtual __exec_system_scheduler_interface* get_completion_scheduler() noexcept = 0;
+};
+
+
+
+
+
+
+// Low-level APIs
+// Phase 2 will move to pointers and ref counting ala COM
+// Phase 3 will move these to weak symbols and allow replacement in tests
+// Default implementation based on static_thread_pool
```

Comment:
Are phases 2 and 3 planned for this PR, or will they happen in a future PR?

Reply:
In a future PR. I'll try to keep the PRs small and focused, and this one is already pretty big.